File: src/main/java/com/mobilesolutionworks/android/zero/CredentialService.java
package com.mobilesolutionworks.android.zero;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by yunarta on 5/5/17.
 */

public class CredentialService {

    private Map<Class<? extends CredentialManager>, CredentialManager> credentialManagerMap;

    CredentialService() {
        this.credentialManagerMap = new HashMap<>();
    }

    <T extends CredentialManager> void installCredentialManager(Class<T> name, T credentialInterface) {
        credentialManagerMap.put(name, credentialInterface);
    }

    public <T extends CredentialManager> T get(Class<T> name) {
        return (T) credentialManagerMap.get(name);
    }
}

File: src/main/kotlin/com/mobilesolutionworks/android/struts/CredentialService.kt
package com.mobilesolutionworks.android.struts

/**
 * Created by yunarta on 13/5/17.
 */

public class CredentialService internal constructor() {

    private val credentialManagerMap: MutableMap<Class<out CredentialManager<*>>, CredentialManager<*>>

    init {
        this.credentialManagerMap = java.util.HashMap<Class<out CredentialManager<*>>, CredentialManager<*>>()
    }

    internal fun <T : CredentialManager<*>> installCredentialManager(name: Class<T>, credentialInterface: T) {
        credentialManagerMap.put(name, credentialInterface)
    }

    @Suppress("UNCHECKED_CAST")
    operator fun <T : CredentialManager<*>> get(name: Class<T>): T? = credentialManagerMap[name] as? T
}
